### DISCUSSION TITLE
单机1.0版本入库每隔4.5h左右出现CPU利用率飙高

### DISCUSSION THREAD

**Discussion by ToughWay** [milvus-io/milvus#17555] [Q&A and General discussion]
单机部署1.0版本持续插入3000多万条向量数据，CPU利用率大部分时间比较低（8核机器cpu占用率2%左右），但是每隔4.5h左右会出现3～5分钟的CPU利用率飙高（8核机器cpu 20%），麻烦问下有什么办法可以确认这个现象的原因么？谢谢。索引是IVFLAT。
![image](https://user-images.githubusercontent.com/63902539/173730144-aad310bc-ec58-41fc-aec7-fce86430cb5a.png)

**Comment by yhmo** [comment]
是sqlite还是mysql？建表后有没有调用create_index定义索引?
最大的可能性是后台在建索引。如果没有调用过create_index，那可能的原因的后台在做比较重的compaction操作(合并多个较大的segment)

**Reply by ToughWay** [reply]
哦，感谢回复。
使用的是阿里云官网的RDS，类似于mysql。
使用方式是先创建表，创建表之后就调用创建索引create_index了，然后才开始持续不断插入向量数据。

**Reply by ToughWay** [reply]
如果猜测是compaction操作，会有相关日志记录么？询问这个CPU利用率问题主要是想评估下milvus需要的机型配置，毕竟大部分时间CPU利用率不高想看下是否调整单机配置。

**Reply by yhmo** [reply]
创建表之后就调用创建索引create_index的话，当插入数据达到一定量(index_file_size)时，会触发后台建立索引，建立索引时消耗的cpu是蛮高的。
在持续insert的过程中，小的segment会不断合并成大的segment，在接近index_file_size这个大小时，合并的操作会比较重。合并时log中会出现 FilesToMerge: SELECT id, table_id, segment_id, file_id, file_type, file_size, row_count, date, engine_type, created_on, updated_time ...... 这样的信息

**Reply by ToughWay** [reply]
哦哦，看了下创建表的时候指定了index_file_size参数，是256MB。
1、当插入数据达到一定量(index_file_size)时，会触发后台建立索引；第一次创建索引成功后，如果继续insert向量，后面的向量是会立即构建索引，还是等后面插入的向量数据再次达到index_file_size时，然后再次触发后台建立索引么？
2、FilesToMerge的日志蛮多的，基本上每1s1条日志。。
![image](https://user-images.githubusercontent.com/63902539/173796918-d1ae3d3e-2919-4855-9181-92112def8fa0.png)

3、再麻烦问下这里的file_type是指？看大约有1、3、7这三个，file_type为1是指原始向量待合并么？
![image](https://user-images.githubusercontent.com/63902539/173793052-e1171f3d-4b14-4ebe-a5b5-7cbe237abae5.png)
4、持久化的原始向量和存储文件可以建议下使用什么软件可以查看内容么？想直观看下。
![image](https://user-images.githubusercontent.com/63902539/173795249-3cbf2a5a-3afd-4953-a981-49a95d67f643.png)
谢谢

**Reply by yhmo** [reply]
插入的向量数据先是放在缓存里，定时flush机制会把缓存里的数据刷到硬盘上，形成一个小分片。后台会把小分片合并，逐渐合并成大分片。1MB和1MB的小分片合并速度很快，但几百兆的大分片合并时会消耗一些机器资源包括内存cpu。当合并后的分片达到index_file_size，这个分片就不再参与合并，后台的线程检测到这个大分片，就开始给它建索引。

FilesToMerge很多，是因为一直不断地有小分片产生并参与合并。

file_type   1是指原始向量文件，3是指索引文件。7其实和1是一个东西，当索引建立好后，1就改为7，表示这个原始向量文件作为备份保存。

原始向量文件和索引文件都是二进制的，没有软件能查看。

**Reply by ToughWay** [reply]
哦哦，非常感谢。
